File: notification-engine-spring/backend/src/main/java/com/nexus/notification/service/DedupeService.java
package com.nexus.notification.service;

import com.nexus.notification.model.Notification;
import com.nexus.notification.repository.NotificationRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import java.time.LocalDateTime;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

@Service
@RequiredArgsConstructor
public class DedupeService {

    private final NotificationRepository notificationRepository;

    public boolean isExactDuplicate(String dedupeKey) {
        return notificationRepository.findByDedupeKey(dedupeKey).isPresent();
    }

    public boolean isNearDuplicate(String userId, String message) {
        LocalDateTime fifteenMinsAgo = LocalDateTime.now().minusMinutes(15);
        List<Notification> recentNotifications = notificationRepository.findByUserIdAndTimestampAfter(userId,
                fifteenMinsAgo);

        for (Notification notif : recentNotifications) {
            if (calculateSimilarity(message, notif.getMessage()) > 0.8) {
                return true;
            }
        }
        return false;
    }

    private double calculateSimilarity(String str1, String str2) {
        Set<String> set1 = new HashSet<>(Arrays.asList(str1.toLowerCase().split("\\s+")));
        Set<String> set2 = new HashSet<>(Arrays.asList(str2.toLowerCase().split("\\s+")));

        Set<String> intersection = new HashSet<>(set1);
        intersection.retainAll(set2);

        Set<String> union = new HashSet<>(set1);
        union.addAll(set2);

        if (union.isEmpty())
            return 0.0;
        return (double) intersection.size() / union.size();
    }
}

File: notification-engine-mern/backend/src/services/DedupeService.ts
import { Notification } from '../models/Notification';

export class DedupeService {
    /**
     * Checks for exact duplicate using the dedupe_key.
     * Returns true if an exact duplicate exists.
     */
    public static async isExactDuplicate(dedupe_key: string): Promise<boolean> {
        const existing = await Notification.findOne({ dedupe_key });
        return !!existing;
    }

    /**
     * Naive near-duplicate detection: Checks if the user sent a VERY similar message
     * within the last 15 minutes. Simple Levenshtein distance could be used, but for
     * basic DB efficiency, we'll check recent messages and do a basic inclusion check.
     */
    public static async isNearDuplicate(userId: string, message: string): Promise<boolean> {
        const fifteenMinsAgo = new Date(Date.now() - 15 * 60 * 1000);

        // Find recent messages by user
        const recentNotifications = await Notification.find({
            user_id: userId,
            timestamp: { $gte: fifteenMinsAgo }
        }).select('message');

        // Simple near-duplicate heuristic:
        // If an existing recent message is an exact substring, or >80% similar, consider it near duplicate.
        for (const notif of recentNotifications) {
            if (this.calculateSimilarity(message, notif.message) > 0.8) {
                return true;
            }
        }
        return false;
    }

    // Very basic Jaccard similarity for words
    private static calculateSimilarity(str1: string, str2: string): number {
        const set1 = new Set(str1.toLowerCase().split(' '));
        const set2 = new Set(str2.toLowerCase().split(' '));
        const intersection = new Set([...set1].filter(x => set2.has(x)));
        const union = new Set([...set1, ...set2]);
        return intersection.size / union.size;
    }
}
